Provide loading and saving functions that use STL streams + preserve extra binary chunks per glTF 2.0 spec

skrat opened the pull request:
The reasons for this are outlined in #59

jessey-git commented on include/fx/gltf.h:
```diff
@@ -798,6 +805,7 @@ namespace gltf
         int32_t scene{ -1 };
         std::vector<std::string> extensionsUsed{};
         std::vector<std::string> extensionsRequired{};
+        std::vector<detail::Chunk> extraChunks;
```
The Document is a public facing object so it having a container of detail::Chunks seems wrong - public users shouldn't care about detail:: stuff.  Moreover the Chunks are pretty low level and I imagine many implementations will want to take the raw byte buffer and map it to their own structures(??).

Would a different design be better?  What if the file load/save paths had some extensibility points, like callbacks, that implementations could opt-in to?  For example, when loading, their callback would be called with information that points to the extra chunks and then they are free to process it directly instead of waiting for it to be packed into these Chunks here?

jessey-git commented on include/fx/gltf.h:
```diff
@@ -1697,6 +1708,20 @@ namespace gltf
                     std::memcpy(&buffer.data[0], &binary[HeaderEndOffset], buffer.byteLength);
                 }
             }
+            if (dataContext.extraChunksLength != 0)
```
In general, leave a blank line after each closing }.  This goes for several other places in the code as it seems clang-format doesn't have an option to enforce this...

jessey-git commented on include/fx/gltf.h:
```diff
@@ -1757,46 +1782,57 @@ namespace gltf
+                for(auto & chunk : document.extraChunks)
+                {
+                    os.write(reinterpret_cast<char const *>(&chunk.header), detail::ChunkHeaderSize);
+                    os.write(reinterpret_cast<char const *>(&chunk.body[0]), chunk.header.chunkLength);
```
Per the spec, each Chunk needs to be aligned on a 4-byte boundary so that needs to be accounted for here.  I have a question out to the spec folks about what should be used as the padding bytes though.

jessey-git commented on include/fx/gltf.h:
```diff
@@ -1806,24 +1842,27 @@ namespace gltf
+        inline void Save(Document const & document, std::string const & documentFilePath, bool useBinaryFormat)
+        {
+            std::ofstream file(documentFilePath, std::ios::binary);
+            if (!file.good())
```
Should move this check into SaveToStream

jessey-git commented on include/fx/gltf.h:
```diff
@@ -1839,46 +1878,156 @@ namespace gltf
+        {
+            detail::GLBHeader header;
+            std::size_t totalSize = 0;
+            std::vector<uint8_t> binary(detail::HeaderSize);
```
Read into a local var first so we don't have to resize the binary vector later (it's a small size but still best to avoid where possible)

jessey-git commented on include/fx/gltf.h:
```diff
@@ -1839,46 +1878,156 @@ namespace gltf
+                    {
+                        throw invalid_gltf_document("Quota exceeded : file size > MaxFileSize");
+                    }
+                    binary.resize(totalSize);
```
This resize will also be pretty expensive...  Is it possible to read from the stream fewer times as well as make the sanity checks against the various sizes and limits just once?  This code path was already pretty tricky, fuzzing found some security issues, and now it's even more complex with all the offsets and additions.

jessey-git commented on include/fx/gltf.h:
```diff
@@ -1839,46 +1878,156 @@ namespace gltf
+        try
+        {
+            std::ifstream file(documentFilePath, std::ios::binary);
+            if (!file.is_open())
```
This check should be moved into LoadFromBinaryStream.  Then the try/catches here can all be removed and let LoadFromBinaryStream handle everything.

jessey-git commented on test/src/unit-exceptions.cpp:
```diff
@@ -141,10 +141,10 @@ TEST_CASE("exceptions")
         fx::gltf::ReadQuotas readQuotas{};
         readQuotas.MaxBufferByteLength = 500;
-        REQUIRE_THROWS_MATCHES(fx::gltf::LoadFromText(externalFile, readQuotas), fx::gltf::invalid_gltf_document, ExceptionContainsMatcher("MaxBufferByteLength"));
+        REQUIRE_THROWS_MATCHES(fx::gltf::LoadFromText(externalFile, readQuotas), fx::gltf::invalid_gltf_document, ExceptionContainsMatcher("MaxBufferByteLength", true));
```
If at all possible I'd like for the exceptions to not be nested.  c++ exceptions are already rather weird and I'd hate to all 3 of these now nested.

skrat commented:
@jessey-git Thanks for the feedback, I'll be looking into it asap.

skrat commented:
Hi @jessey-git , I moved the chunks commits elsewhere for now, I updated the code according to the comments. If it's all right, and merged in the master, I'll make another PR just for chunks. Does that sound reasonable?

skrat commented:
Hi @jessey-git just pinging back if all is well and if there are any plans to merge just the iostream stuff for now. Thank you

jessey-git commented:
Haven't forgotten this but have been busy.  I finally got some time to look into this further to make some changes.  I put everything in the `iostream` branch here.  I started with your patch and did the following:
- Used overloads for the various Load/Save functions
- Removed the need to resize the big .glb blob by splitting into 2 blobs of exact size instead
- Edited some whitespace
- Added tests for the new iostream support

Can you give that branch a look over to see if it still meets your expectations?

skrat commented:
@jessey-git perfect, meets all expectations.

jessey-git commented:
Closed by ways of another merge
